### PR TITLE
React: refactor the way relations are handled

### LINKED ISSUE #7967: React: saving failed when using relationship to Jhipster User
##### **Overview of the issue**

I am using example from https://github.com/mraible/jhipster5-demo. I used Postgresql for dev and pro. Project with entities are generated successfully.

When I create new Blog and saving, i got this error.

Exception in com.asdev.blog.web.rest.BlogResource.createBlog() with cause = 'java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.asdev.blog.domain.Blog.user -> com.asdev.blog.domain.User' and exception = 'org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.asdev.blog.domain.Blog.user -> com.asdev.blog.domain.User; nested exception is java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.asdev.blog.domain.Blog.user -> com.asdev.blog.domain.User'

##### **Motivation for or Use Case**

I used previous jhipster 4* and worked

##### **Reproduce the error**

Run the example then create new Blog 

##### **JHipster Version(s)**

```
blog@0.0.0 /Volumes/Data/*
└── generator-jhipster@5.1.0 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.asdev.indag.umkm"
    },
    "jhipsterVersion": "5.1.0",
    "applicationType": "monolith",
    "baseName": "blog",
    "packageName": "com.asdev.indag.umkm",
    "packageFolder": "com/asdev/indag/umkm",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "react",
    "useSass": false,
    "clientPackageManager": "yarn",
    "testFrameworks": [
      "protractor"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": false
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Blog (blog) {
  name String required minlength(3),
  handle String required minlength(2)
}
entity Entry (entry) {
  title String required,
  content TextBlob required,
  date Instant required
}
entity Tag (tag) {
  name String required minlength(2)
}
relationship ManyToOne {
  Blog{user(login)} to User,
  Entry{blog(name)} to Blog
}
relationship ManyToMany {
  Entry{tag(name)} to Tag{entry}
}

paginate Entry, Tag with infinite-scroll

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_112"
Java(TM) SE Runtime Environment (build 1.8.0_112-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.112-b16, mixed mode)

git version 2.15.2 (Apple Git-101.1)

node: v10.6.0

npm: 6.1.0

yeoman: 2.0.2

yarn: 1.7.0

##### **Browsers and Operating System**

OSX 10.13.5, Safari, Chrome

- [X] Checking this box is mandatory (this is just to show you read everything)

### LINKED ISSUE #7986: React - id value not present when linking through field on a relationship using JDL -> Entity{fk(field)}
##### **Overview of the issue**

No id present when updating an entity if the relationship was defined to be linked through field different from default(the ID)
i.e.(using name):
`relationship OneToMany {
  Author{book(name)} to Book{writer(name)}
}`
##### **Motivation for or Use Case**
Updating an existing value on DB.

Server will throw and exception since the id of the relationship(which exists) is not present. Even if you are changing other fields(leaving the relationship as it is)
##### **Reproduce the error**
Update a value on an existing record;

TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing

##### **Related issues**
Maybe: https://github.com/jhipster/generator-jhipster/issues/7967#issuecomment-405064024

`<AvInput id="blog-user" type="hidden" name="user.id" value={this.state.userId} />
`
5.1.0 did not generated above code. I added the code in 5.1.0 and works perfectly.

##### **JHipster Version(s)**
5.1.0

##### **JHipster configuration**

```
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.test",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.1.0",
    "applicationType": "gateway",
    "baseName": "PTasks",
    "packageName": "com.test",
    "packageFolder": "com/test",
    "serverPort": "9980",
    "authenticationType": "uaa",
    "uaaBaseName": "uaa",
    "cacheProvider": "hazelcast",
    "enableHibernateCache": true,
    "websocket": "spring-websocket",
    "databaseType": "sql",
    "devDatabaseType": "mysql",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": "eureka",
    "buildTool": "gradle",
    "enableSwaggerCodegen": false,
    "clientFramework": "react",
    "useSass": false,
    "clientPackageManager": "yarn",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "es"
    ]
  }
}
```
##### **Browsers and Operating System**

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by agaspardcilia**
fix #7967 
fix #7986  

- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by deepu105** [comment]
@jhipster/developers can someone test this please

File: generators/client/templates/react/src/test/javascript/spec/app/shared/util/entity-utils.spec.ts.ejs [hunk 1]
```diff
@@ -59,4 +59,18 @@ describe('Entity utils', () => {
       expect(cleanEntity({ ...entity })).toEqual(entity);
     });
   });
+
+  describe('formatIdList', () => {
```
Comment (deepu105) [inline generators/client/templates/react/src/test/javascript/spec/app/shared/util/entity-utils.spec.ts.ejs]: change to correct method name

**Comment by deepu105** [comment]
@ruddell will you be able to test this?

**Comment by ruddell** [comment]
Looks good, still fails if you select the blank item at the top of a ManyToMany dropdown.  Simple fix is to filter for empty items before mapping:

    export const mapIdList = (idList: ReadonlyArray<any>) => idList.filter((entityId: any) => entityId !== '').map((entityId: any) => ({ id: entityId }));

**Comment by MathieuAA** [comment]
I don't like the string equality sign... It introduces type dependency... do you think there's a simple way to fix that? Like a clear functionLe 3 août 2018 21:08, Jon Ruddell <notifications@github.com> a écrit :Looks good, still fails if you select the blank item at the top of a ManyToMany dropdown..n.  Simple fix is to filter for empty items before mapping:
export const mapIdList = (idList: ReadonlyArray<any>) => idList.filter((entityId: any) => entityId !== '').map((entityId: any) => ({ id: entityId }));

**Comment by deepu105** [comment]
@MathieuAA didnt get you, could you point to the line you are referring to?

**Comment by MathieuAA** [comment]
idList.filter((entityId: any) => entityId !== '')I mean, why are there null IDs in the first place?

**Comment by ruddell** [comment]
There's a blank option at the top for selecting "none" of the entities

**Comment by delkant** [comment]
After generating some entities with JDL and the latest code in master, when building a React client.. I am getting an error (listed below) for all entities on the *-update.tsx files. I think it might be related to this change.

I couldn't fine any reference to keysToValues in the code so I am not sure where is this coming from.

**Error:**
ERROR in $APP_HOME/src/main/webapp/app/entities/item/$ENTITY_NAME-update.tsx
Module '"$APP_HOME/src/main/webapp/app/shared/util/entity-utils"' has no exported member 'keysToValues'.

keysToValues was remove in entity-utils.tsx in one of these commits.

**Comment by agaspardcilia** [comment]
@delkant I had a similar problem while on that PR. It was due to Yarn going haywire after a `yarn link generator-jhipster` (it's yarn bug that is supposed to be fixed in its next release). I mean you're very unlikely to have issues with `keysToValues` in the master since there's not a single reference to this function anymore in the whole project.

If you're using the last master make sure it's up to date and that if you're using yarn make sure it's not the problem by trying to delete `yarn.lock` and `node_modules` and running a `yarn install` on the generator.

**Pull request merged**